boot/boot.go

```
@@ -0,0 +1,171 @@
+)
+
+// A Device Boot Set is a collection of snaps that are involved in a
+// device's boot process.
+type DeviceBootSet interface {
+	// RemoveKernelAssets removes the unpacked kernel/initrd for the given
+	// kernel snap.
+	RemoveKernelAssets() error
+	// ExtractKernelAssets extracts kernel/initrd/dtb data from the given
+	// kernel snap, if required, to a versioned bootloader directory so
+	// that the bootloader can use it.
+	ExtractKernelAssets(snap.Container) error
+	// SetNextBoot will schedule the given OS or base or kernel snap to be
+	// used in the next boot. For base snaps it up to the caller to select
+	// the right bootable base (from the model assertion).
+	SetNextBoot() error
+	// ChangeRequiresReboot returns whether a reboot is required to switch
+	// to the given OS, base or kernel snap.
+	ChangeRequiresReboot() bool
+}
+
+// Lookup figures out what the right boot set is for the given
+// arguments, and returns it.
+func Lookup(s snap.PlaceInfo, t snap.Type, onClassic bool) (DeviceBootSet, error) {
+	if t != snap.TypeOS && t != snap.TypeKernel && t != snap.TypeBase && t != snap.TypeSnapd {
+		return nil, fmt.Errorf("cannot lookup boot set with snap %q of type %q", s.SnapName(), t)
+	}
+
+	if onClassic {
+		return classicBootSet{}, nil
+	}
+
+	return &coreBootSet{s: s, t: t}, nil
+}
+
+type classicBootSet struct{}
+
+func (classicBootSet) RemoveKernelAssets() error {
+	return fmt.Errorf("cannot remove kernel assets on classic systems")
+}
+
+func (classicBootSet) ExtractKernelAssets(snap.Container) error {
+	return fmt.Errorf("cannot extract kernel assets on classic systems")
+}
+
+func (classicBootSet) SetNextBoot() error {
+	return fmt.Errorf("cannot set next boot on classic systems")
+}
+
+func (classicBootSet) ChangeRequiresReboot() bool { return false }
+
+var _ DeviceBootSet = classicBootSet{}
+
+// InUse checks if the given name/revision is used in the
+// boot environment
+func InUse(name string, rev snap.Revision) bool {
```

I would be more comfortable with a function that returns an error rather than logging around. It's hard to determine if this is the right layer for ignoring + ignoring errors vs bubbling them up.

---

this is not new code in this PR